google/cloud/forseti/scanner/audit/enabled_apis_rules_engine.py

```diff
@@ -0,0 +1,401 @@
+
+LOGGER = logger.get_logger(__name__)
+
+# Rule Modes.
+_WHITELIST = 'whitelist'
+_BLACKLIST = 'blacklist'
+_REQUIRED = 'required'
+_RULE_MODES = frozenset([_WHITELIST, _BLACKLIST, _REQUIRED])
+
+# Rule violation types.
+_VIOLATION_TYPES = {
+    _WHITELIST: 'ENABLED_APIS_ADDED_VIOLATION',
+    _BLACKLIST: 'ENABLED_APIS_ADDED_VIOLATION',
+    _REQUIRED: 'ENABLED_APIS_REMOVED_VIOLATION',
+}
+
+
+def _check_whitelist_apis(rule_apis=None, enabled_apis=None):
+    """Whitelist: Checks that enabled APIs ARE in rule APIs.
+
+    If an enabled api is NOT found in the rule APIs, add it to
+    the violating APIs.
+
+    Args:
+        rule_apis (list): ManagedService APIs allowed in the rule.
+        enabled_apis (list): ManagedService APIs enabled by the project.
+
+    Returns:
+        list: Enabled APIs NOT found in the whitelist (rule APIs).
+    """
+    return [api for api in enabled_apis if api not in rule_apis]
+
+
+def _check_blacklist_apis(rule_apis=None, enabled_apis=None):
+    """Blacklist: Checks that enabled APIs ARE NOT in rule APIs.
+
+    If an enabled API is found in the rule APIs, add it to the
+    violating APIs.
+
+    Args:
+        rule_apis (list): ManagedService APIs dis-allowed in the rule.
+        enabled_apis (list): ManagedService APIs enabled by the project.
+
+    Returns:
+        list: Enabled APIs found in the blacklist (rule APIs).
+    """
+    return [api for api in enabled_apis if api in rule_apis]
+
+
+def _check_required_apis(rule_apis=None, enabled_apis=None):
+    """Required: Checks that rule APIs are in enabled APIs.
+
+    If a required rule API is NOT found enabled in the project, add it to the
+    violating APIs.
+
+    Args:
+        rule_apis (list): ManagedService APIs required in the rule.
+        enabled_apis (list): ManagedService APIs enabled by the project.
+
+    Returns:
+        list: Required rule APIs not found enabled by the project.
+    """
+    return [api for api in rule_apis if api not in enabled_apis]
+
+
+class EnabledApisRulesEngine(bre.BaseRulesEngine):
+    """Rules engine for enabled APIs."""
+
+    def __init__(self, rules_file_path, snapshot_timestamp=None):
+        """Initialize.
+
+        Args:
+            rules_file_path (str): File location of rules.
+            snapshot_timestamp (str): The snapshot to work with.
+        """
+        super(EnabledApisRulesEngine, self).__init__(
+            rules_file_path=rules_file_path,
+            snapshot_timestamp=snapshot_timestamp)
+        self.rule_book = None
+
+    def build_rule_book(self, global_configs=None):
+        """Build EnabledApisRuleBook from the rules definition file.
+
+        Args:
+            global_configs (dict): Global configurations.
+        """
+        self.rule_book = EnabledApisRuleBook(
+            global_configs,
+            self._load_rule_definitions(),
+            snapshot_timestamp=self.snapshot_timestamp)
+
+    def find_policy_violations(self, project, enabled_apis,
```

Comment:
Really small nit: can you please change this method name to just `find_violations()`?  i.e. remove `policy`

Reply:
Fixed.